### PR TITLE
Feature/show studs in assembly view

### LINKED ISSUE #669: Studs that are not occluded are hidden anyways
in Lego assembly view.
image

### PR DESCRIPTION
**Pull request by stuikomma**
fixes #669

### CODE REVIEW

File: src/plugins/newBrickator/pipeline/Brick.coffee [hunk 1]
```diff
@@ -182,16 +182,17 @@ class Brick
 		@_neighbors = null
 		@_isCoveredOnTop = null
 
-	# Returns true if the brick is fully covered by other bricks on its top,
-	# or false, if there is (partly) 3D-print-geometry above it
+	# Returns the covering bricks if the brick is fully covered by other bricks
+	# on its top, or null, if there is (partly) 3D-print-geometry above it
 	isCoveredOnTop: =>
```
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/Brick.coffee]: Please rename - call it `getCover` or similar - right now it is too easy confusable with `@_isCoveredOnTop`, but returns totally different things.

File: src/plugins/newBrickator/pipeline/Brick.coffee [hunk 2]
```diff
@@ -182,16 +182,17 @@ class Brick
...
 		if @_isCoveredOnTop?
-			return @_isCoveredOnTop
+			return if @_isCoveredOnTop then @getNeighbors Brick.direction.Zp else null
```
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/Brick.coffee]: Maybe always return
```js
{
   cover: @getNeighbors [...]
   isCompletelyCovered: true/false
}
```

Comment (Steditor) [inline src/plugins/newBrickator/pipeline/Brick.coffee]: The `if @_isCoveredOnTop` seems redundant...

File: src/plugins/nodeVisualizer/visualization/brickVisualization.coffee [hunk 3]
```diff
@@ -156,7 +159,15 @@ class BrickVisualization
 		@_visibleChildLayers = null
 
 	_setStudVisibility: (brick) ->
-		brick.getVisualBrick().setStudVisibility not brick.isCoveredOnTop()
+		result = brick.isCoveredOnTop()
```
Comment (Steditor) [inline src/plugins/nodeVisualizer/visualization/brickVisualization.coffee]: Make the variable name more descriptive than `result`

File: src/plugins/nodeVisualizer/visualization/brickVisualization.coffee [hunk 4]
```diff
@@ -179,16 +190,28 @@ class BrickVisualization
...
+					@_makeLayerColored threeLayer
+					for visibleBrick in threeLayer.children
+						visibleBrick.visible = true
+						visibleBrick.setStudVisibility true
 			else
-				threeLayer.visible = false
+				for visibleBrick in threeLayer.children
+					visibleBrick.visible = false
+
+		# Set stud visibility in second pass so that visibility of
```
Comment (Steditor) [inline src/plugins/nodeVisualizer/visualization/brickVisualization.coffee]: The comment does not really explain, why a second pass is necessary - obviously we are doing it to make something correct, but what?

File: src/plugins/newBrickator/pipeline/Brick.coffee [hunk 5]
```diff
@@ -182,16 +182,17 @@ class Brick
...
-	isCoveredOnTop: =>
-		if @_isCoveredOnTop?
-			return @_isCoveredOnTop
-
-		stability = @fractionOfConnectionsInZDirection Brick.direction.Zp
-
-		@_isCoveredOnTop = stability > 0.99
-		return @_isCoveredOnTop
+	# Returns whether the brick is completely covered by other bricks
+	# (isCompletelyCovered) and a set of covering bricks (coveringBricks)
```
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/Brick.coffee]: Maybe use the formal documentation with `@` annotations?

**Comment by Steditor** [comment]
Do we want to put it in v0.10.2 as hotfix as well? Then merge before #673

**Comment by stuikomma** [comment]
I agree that it should be merged before v0.10.2

**Comment by Steditor** [comment]
I'm done with a rough code review - s.o. else needs to test this, as I don't have my setup here.

**Comment by AtiX** [comment]
:rocket:

**Pull request merged**